### PR TITLE
introduced DelegatingFileLoader to load different webspace configuration formats

### PR DESCRIPTION
**Pull request by danrot**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | yes
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | https://github.com/sulu/sulu-standard/pull/727 https://github.com/sulu/sulu-minimal/pull/9 https://github.com/sulu/sulu-docs/pull/240/files
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR splits the `XmlFileLoader` for the webspaces in two different ones, to allow loading different webspace configuration formats (e.g. using version 1.0 or 1.1 of the webspace schema). It also changes the `error-templates` tag to a simple `templates` tag, so that predefined templates can also be used for other stuff than errors.

#### Why?

This PR is a prerequisite to move the `SearchController` from sulu-standard to sulu. It allows to add a `template` tag typed with `search`, so that the `SearchController` can take this value to render the search result.

#### Example Usage

The webspace.xml has to contain tags like this:

~~~xml
<templates>
    <template type="error-404">SomeBundle:views:error404.html.twig</template>
    <template type="error">SomeBundle:views:error.html.twig</template>
</templates>
~~~

instead of these:

~~~xml
<error-templates>
    <error-template code="404">SomeBundle:view:error404.html.twig</error-template>
    <error-template default="true">SomeBundle:view:error.html.twig</error-template>
</error-templates>
~~~

#### BC Breaks/Deprecations

The `error-template` tags in the webspace configuration do not exist anymore, and the `XmlFileLoader` structure for the webspaces changed, but it's unlikely that somebody was extending them.

#### To Do

- [x] Create a documentation PR
- [x] Add tests for `ExceptionController`
- [x] Add changelog

### CODE REVIEW

File: src/Sulu/Component/Webspace/Loader/BaseXmlFileLoader.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function supports($resource, $type = null)
+    {
+        if (!is_string($resource) || 'xml' !== pathinfo($resource, PATHINFO_EXTENSION)) {
+            return false;
+        }
+
+        $document = XmlUtils::loadFile($resource);
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Loader/BaseXmlFileLoader.php]: Haven't found any fast better solution to do this. The file has to be read to know which loader should be used, and caching the file would mean that this service introduces state, which should be avoided. Also, this operation is only called when writing the cache, so it shouldn't be a big deal anyway.

File: UPGRADE.md [hunk 2]
```diff
@@ -1,5 +1,30 @@
 # Upgrade
 
+## dev-develop
+
+### Webspace Configuration
+
+The configuration schema for webspaces has changed. Instead of
```
Comment (wachterjohannes) [inline UPGRADE.md]: you should mention that this only counts when your using 1.1 schema

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader11.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Webspace\Loader;
+
+use Sulu\Component\Webspace\Loader\Exception\ExpectedDefaultTemplatesNotFound;
+use Sulu\Component\Webspace\Webspace;
+
+class XmlFileLoader11 extends XmlFileLoader10
```
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader11.php]: phpdoc

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader11.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+    const SCHEMA_LOCATION = '/schema/webspace/webspace-1.1.xsd';
+
+    const SCHEMA_URI = 'http://schemas.sulu.io/webspace/webspace-1.1.xsd';
+
+    protected function generateDefaultTemplates(Webspace $webspace)
```
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader11.php]: phpdoc

File: src/Sulu/Component/Webspace/Loader/XmlFileLoader11.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        foreach ($expected as $item) {
+            if (!in_array($item, $found)) {
+                throw new ExpectedDefaultTemplatesNotFound($this->webspace->getKey(), $expected, $found);
+            }
+        }
+
+        return $webspace;
+    }
+
+    protected function generateTemplates(Webspace $webspace)
```
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Loader/XmlFileLoader11.php]: phpdoc

**Comment by wachterjohannes** [comment]
@danrot ACK please fix this minor issues

**Pull request merged**